tests/test_image_untagger.py

```diff
@@ -0,0 +1,349 @@
+from .utils.misc import sort_dictionary_sortable_values, compare_logs
+
+# flake8: noqa: E501
+
+
+@mock.patch("pubtools._quay.image_untagger.QuayClient")
+@mock.patch("pubtools._quay.image_untagger.QuayApiClient")
+def test_init_success(mock_quay_api_client, mock_quay_client):
+    references = ["stage.quay.io/name/repo1:1", "stage.quay.io/name/repo2:2"]
+    token = "some-token"
+    quay_user = "some-user"
+    quay_password = "some-password"
+    host = "stage.quay.io/"
+    untagger = image_untagger.ImageUntagger(
+        references, token, False, quay_user, quay_password, host
+    )
+
+    assert untagger.references == references
+    assert untagger.host == host[:-1]
+    assert not untagger.remove_last
+    mock_quay_client.assert_called_once_with(quay_user, quay_password, host[:-1])
+    mock_quay_api_client.assert_called_once_with(token, host[:-1])
+
+
+@mock.patch("pubtools._quay.image_untagger.QuayClient")
+@mock.patch("pubtools._quay.image_untagger.QuayApiClient")
+def test_init_bad_reference(mock_quay_api_client, mock_quay_client):
+    references = [
+        "stage.quay.io/name/repo1:1",
+        "stage.quay.io/name/repo2@sha256:dgdgdg",
+    ]
+    token = "some-token"
+    quay_user = "some-user"
+    quay_password = "some-password"
+    host = "stage.quay.io/"
+    with pytest.raises(ValueError, match=".*must be specified via tag, not digest.*"):
+        image_untagger.ImageUntagger(
+            references, token, False, quay_user, quay_password, host
+        )
+
+
+@mock.patch("pubtools._quay.image_untagger.QuayClient")
+@mock.patch("pubtools._quay.image_untagger.QuayApiClient")
+def test_init_no_docker_client(mock_quay_api_client, mock_quay_client):
+    references = ["stage.quay.io/name/repo1:1", "stage.quay.io/name/repo2:2"]
+    token = "some-token"
+    quay_user = "some-user"
+    host = "stage.quay.io/"
+    untagger = image_untagger.ImageUntagger(
+        references, token, False, quay_user, host=host
+    )
+
+    assert untagger.references == references
+    assert untagger.host == host[:-1]
+    assert not untagger.remove_last
+    mock_quay_client.assert_not_called()
+    mock_quay_api_client.assert_called_once_with(token, host[:-1])
+    assert untagger._quay_client is None
+
+    client = quay_client.QuayClient("user", "pass")
+    untagger.set_quay_client(client)
+    assert client == untagger._quay_client
+
+
+@mock.patch("pubtools._quay.image_untagger.QuayClient")
+@mock.patch("pubtools._quay.image_untagger.QuayApiClient")
+def test_repo_tag_mapping(mock_quay_api_client, mock_quay_client):
+    references = [
+        "stage.quay.io/name/repo1:1",
+        "stage.quay.io/name/repo1:2",
+        "stage.quay.io/name/repo2:2",
+        "stage.quay.io/name2/repo1:1",
+        "stage.quay.io/name2/repo1:3",
+    ]
+    token = "some-token"
+    quay_user = "some-user"
+    quay_password = "some-password"
+    host = "stage.quay.io/"
+    untagger = image_untagger.ImageUntagger(
```

Comment:
can you make this into fixture?

Reply:
Not sure I understand. Why can't I just invoke the class directly?

Reply:
Of course you can but lines 82-88 are repeating over several tests. So to make test cases smaller it make sense to put it in fixture

Reply:
But the reference parameter is almost always different. How am I supposed to pass it to the fixture?

Reply:
yeah, that would be ugly. So just some setup_untagger(references) function could help as well

Reply:
References change in almost every test but other options changes as well sometimes. I think it will be more confusing if this setup will only be used in some specific cases instead of all